### PR TITLE
xds: Clarify NACK detection in protocol spec.

### PR DESCRIPTION
**Pull request by markdroth**
Signed-off-by: Mark D. Roth <roth@google.com>

Commit Message: xds: Clarify NACK detection in protocol spec.
Additional Description: Clarify that NACKs should be detected via the presence of the error_details field, rather than by looking at nonce and version; the latter approach does not work for APIs other than LDS or CDS, where the client can change the set of resources it is subscribing to independently of the resource type instance version changing on the server.  Also clarify semantics of nonce handling.
Risk Level: Low
Testing: N/A
Docs Changes: Included in PR
Release Notes: N/A
Platform Specific Features: N/A

CC @htuch

### CODE REVIEW

**Comment by markdroth** [comment]
For reference, here is an example of where the server cannot detect a NACK by looking at the version and nonce for an API other than LDS and CDS:

- C: sends resource_names=[A]
- S: sends resources=[A], version=1, nonce=1
- C: sends resource_names=[A], version=1, nonce=1 (this is an ACK)
- C: sends resource_names=[A, B], version=1, nonce=1 (client is adding subscription to B)
- S: sends resources=[B], version=1, nonce=2 (server sends B but system version does not change because server has not gotten a new config)
- C: sends resource_names=[A, B], version=1, nonce=2, error_detail="B is invalid" (client intended to NACK B, but server will interpret this as an ACK)

File: api/xds_protocol.rst [hunk 1]
```diff
@@ -356,21 +365,30 @@ After a NACK, an API update may succeed at a new version **Y**:
...
   :ref:`version_info <envoy_api_field_DiscoveryResponse.version_info>` from the
   :ref:`DiscoveryResponse <envoy_api_msg_DiscoveryResponse>`.
-- NACK signifies unsuccessful configuration update and contains the previous (existing)
-  :ref:`version_info <envoy_api_field_DiscoveryResponse.version_info>`.
-- Only the NACK should populate the :ref:`error_detail <envoy_api_field_DiscoveryRequest.error_detail>`.
+- NACK signifies unsuccessful configuration and is indicated by the presence of the
+  :ref:`error_detail <envoy_api_field_DiscoveryRequest.error_detail>` field.  The :ref:`version_info
+  <envoy_api_field_DiscoveryResponse.version_info>` indicates the most recent version that the
+  client is using, although that may not be an older version in the case where the client has
+  subscribed to a new resource from an existing version and that new resource is invalid.
```
Comment (htuch) [inline api/xds_protocol.rst:391]: This last sentence is a bit unclear. I think I get at what you mean based on our earlier conversation but not sure other readers will.

**Review by htuch — CHANGES_REQUESTED** [review]
Appreciate the clarifications!
/wait

File: api/xds_protocol.rst [hunk 2]
```diff
@@ -356,21 +365,30 @@ After a NACK, an API update may succeed at a new version **Y**:
 .. figure:: diagrams/later-ack.svg
    :alt: ACK after NACK
 
+The preferred mechanism for a server to detect a NACK is to look for the presence of the
+:ref:`error_detail <envoy_api_field_DiscoveryRequest.error_detail>` field in the request sent by
+the client. Some older servers may instead detect a NACK by looking at both the version and the
+nonce in the request: if the version in the request is lower than the one sent by the server with
```
[the remark this answers is not in the document]
Comment (markdroth) [reply api/xds_protocol.rst:371]: Fixed.

File: api/xds_protocol.rst (hunk 1 above)
Comment (markdroth) [reply api/xds_protocol.rst:391]: To clarify this, I've added an example above and referred back to it here.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Comment by htuch** [comment]
@markdroth see CI.

**Comment by markdroth** [comment]
The CI failure showing now looks unrelated to the PR.  Harvey, can you (or someone else knowledgeable) please take a look?  Thanks!

**Comment by htuch** [comment]
@markdroth can you merge master and push? Thanks.

**Pull request merged**